libcob/common.c

```diff
@@ -10055,15 +10148,97 @@ cob_call_with_exception_check (const char *name, const int argc, void **argv)
 	return 0;
 }
 
-void
-cob_init (const int argc, char **argv)
+
+static
+void cob_set_main_argv0 (const int argc, char **argv)
 {
 	char		*s;
 #if	defined (HAVE_READLINK) || defined (HAVE_GETEXECNAME)
 	const char	*path;
 #endif
 	int		i;
 
+#ifdef _WIN32
+	s = cob_malloc ((size_t)COB_LARGE_BUFF);
+	i = GetModuleFileNameA (NULL, s, COB_LARGE_MAX);
+	if (i > 0 && i < COB_LARGE_BUFF) {
+		cobglobptr->cob_main_argv0 = cob_strdup (s);
+		cob_free (s);
+		return;
+	}
+	cob_free (s);
+#endif
+#ifdef HAVE_READLINK
+	if (!access ("/proc/self/exe", R_OK)) {
+		path = "/proc/self/exe";
+	} else if (!access ("/proc/curproc/file", R_OK)) {
+		path = "/proc/curproc/file";
+	} else if (!access ("/proc/self/path/a.out", R_OK)) {
+		path = "/proc/self/path/a.out";
+	} else {
+		path = NULL;
+	}
+	if (path) {
+		s = cob_malloc ((size_t)COB_LARGE_BUFF);
+		i = (int)readlink (path, s, (size_t)COB_LARGE_MAX);
+		if (i > 0 && i < COB_LARGE_BUFF) {
+			s[i] = 0;
+			cobglobptr->cob_main_argv0 = cob_strdup (s);
+			cob_free (s);
+			return;
+		}
+		cob_free (s);
+	}
+#endif
+
+#ifdef	HAVE_GETEXECNAME
+	path = getexecname ();
+	if (path) {
+#ifdef	HAVE_REALPATH
+		s = cob_malloc ((size_t)COB_LARGE_BUFF);
+		if (realpath (path, s) != NULL) {
+			cobglobptr->cob_main_argv0 = cob_strdup (s);
+		} else {
+			cobglobptr->cob_main_argv0 = cob_strdup (path);
+		}
+		cob_free (s);
+#else
+		cobglobptr->cob_main_argv0 = cob_strdup (path);
+#endif
+		return;
+	}
+#endif
+
+	if (argc && argv && argv[0]) {
+#if	defined (HAVE_CANONICALIZE_FILE_NAME)
+		/* Returns malloced path or NULL */
+		cobglobptr->cob_main_argv0 = canonicalize_file_name (argv[0]);
+#elif	defined (HAVE_REALPATH)
+		s = cob_malloc ((size_t)COB_LARGE_BUFF);
+		if (realpath (argv[0], s) != NULL) {
+			cobglobptr->cob_main_argv0 = cob_strdup (s);
+		}
+		cob_free (s);
+#elif	defined	(_WIN32)
+		/* Returns malloced path or NULL */
+		cobglobptr->cob_main_argv0 = _fullpath (NULL, argv[0], 1);
+#endif
+		if (!cobglobptr->cob_main_argv0) {
+			cobglobptr->cob_main_argv0 = cob_strdup (argv[0]);
+		}
+	} else {
+		cobglobptr->cob_main_argv0 = cob_strdup (_("unknown"));
+	}
```

Comment:
Looks like we have that code multiple times since quite a while - can you please extract that (either from common.c or from call.c to fileio.c) and call it here (exported in common.c so we can use it in cobc as well?

Sorry for missing that up-front (and yes, I'm aware that this part was just moved - so it is no reason to not mark this PR to need adjustments)...